### PR TITLE
e2e/storage: speed up kubelet commands

### PR DESCRIPTION
**Pull request by huww98**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Speed up stopping by not waiting for Node not ready, `systemctl` will ensure kubelet process stopped before return. This should save 40s per case.

Since stop command does not wait for not ready, start command needs to wait for the next heartbeat to ensure we are checking status from new process.

implement restart by stop then start, to get heartbeat time when kubelet is down. And we do not need to sleep 30s now. The sleep is moved to callers, since they still need them to ensure the volume does not disappear.

Dropped support for non-systemd system.

#### Special notes for your reviewer:

I think the non-systemd setup never worked before. Do a non-systemd system have `Main PID` in its output?

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: test/e2e/framework/node/wait.go [hunk 1]
```diff
@@ -160,6 +160,22 @@ func WaitForNodeSchedulable(ctx context.Context, c clientset.Interface, name str
...
+	for start := time.Now(); time.Since(start) < timeout; time.Sleep(poll) {
+		node, err := c.CoreV1().Nodes().Get(ctx, name, metav1.GetOptions{})
+		if err != nil {
+			framework.Logf("Couldn't get node %s", name)
+			continue
+		}
+		if GetNodeHeartbeatTime(node).After(after.Time) {
+			return true
+		}
+	}
```
Comment (pohly) [inline test/e2e/framework/node/wait.go:174]: Can you use `gomega.Eventually` here instead of writing your own polling code?

Comment (huww98) [reply test/e2e/framework/node/wait.go]: OK, I've also converted two other polling to `gomega.Eventually`

**Comment by huww98** [comment]
@pohly PTAL again, thanks.

File: test/e2e/framework/node/wait.go [hunk 2]
```diff
@@ -116,19 +117,14 @@ func WaitForTotalHealthy(ctx context.Context, c clientset.Interface, timeout tim
...
-		}
-
-		if IsConditionSetAsExpected(node, conditionType, wantTrue) {
-			return true
+			return false, err
 		}
-	}
-	framework.Logf("Node %s didn't reach desired %s condition status (%t) within %v", name, conditionType, wantTrue, timeout)
-	return false
+		return IsConditionSetAsExpected(node, conditionType, wantTrue), nil
```
Comment (pohly) [inline test/e2e/framework/node/wait.go:126]: Go helper functions which return `bool, err` are not very informative because the `false, nil` case comes with no explanation that could be printed in a progress report.

But you are using some existing function here, so it wouldn't be fair to ask you to improve that.

File: test/e2e/framework/node/wait.go [hunk 3]
```diff
@@ -160,32 +156,42 @@ func WaitForNodeSchedulable(ctx context.Context, c clientset.Interface, name str
 	return false
 }
 
+func WaitForNodeHeartbeatAfter(ctx context.Context, c clientset.Interface, name string, after metav1.Time, timeout time.Duration) bool {
```
Comment (pohly) [inline test/e2e/framework/node/wait.go:159]: There's no need for returning the boolean.

The failure handler will abort with a panic, so there's never a situation where this function returns false.

File: test/e2e/storage/utils/utils.go [hunk 4]
```diff
@@ -104,6 +96,9 @@ func TestKubeletRestartsAndRestoresMount(ctx context.Context, c clientset.Interf
 	ginkgo.By("Restarting kubelet")
 	KubeletCommand(ctx, KRestart, c, clientPod)
 
+	ginkgo.By("Wait 20s for the volume to become stable")
+	time.Sleep(20 * time.Second)
```
Comment (pohly) [inline test/e2e/storage/utils/utils.go:100]: Hmm. Magic timeouts have been great (not!) source of test flakes. Can you elaborate on why 20 seconds where chosen? How certain is it that this value works reliably? How does it affect the runtime of tests?

Comment (huww98) [reply test/e2e/storage/utils/utils.go:100]: This sleep is moved from `KubeletCommand()` just above it. We are asserting that something is not happening, so we need a timeout.

I choose 20 seconds because the original code in KubeletCommand():
```go
framework.Logf("Noticed that kubelet PID is changed. Waiting for 30 Seconds for Kubelet to come back")
time.Sleep(30 * time.Second)
```
The sleep now starts after node become ready, so (20s + WaitForNodeToBeReady) should roughly equal to the original 30s.

**Comment by huww98** [comment]
/remove-sig node
/remove-sig cloud-provider
I found the impact of changing WaitConditionToBe() to panic on failure is too wide. Reverted that for this PR.

File: test/e2e/framework/node/wait.go (hunk 2 above)
Comment (huww98) [reply test/e2e/framework/node/wait.go]: Reverted because the impact is too wide.

**Comment by huww98** [comment]
CreateVolume failed to create single zonal disk "pvc-cd42c4cd-45d0-480c-91bc-3357014114f6": failed to insert zonal disk: unknown Insert disk operation error: operation operation-1715853438060-6188f42a98737-9493b122-2ab6b2c5 failed (INTERNAL_ERROR): Internal error. Please try again or contact Google Support. (Code: 'Internal error')

OK, so let me try again

/retest

File: test/e2e/framework/node/wait.go [hunk 5]
```diff
@@ -160,32 +161,38 @@ func WaitForNodeSchedulable(ctx context.Context, c clientset.Interface, name str
 	return false
 }
 
+func WaitForNodeHeartbeatAfter(ctx context.Context, c clientset.Interface, name string, after metav1.Time, timeout time.Duration) {
```
Comment (pohly) [inline test/e2e/framework/node/wait.go:164]: Can you add documentation? Please don't just repeat the function name. Paraphrase it and explain the parameters.

"nodeName" instead of "name" would be more descriptive. "after" is also very generic.

File: test/e2e/framework/node/resource.go [hunk 6]
```diff
@@ -495,6 +495,15 @@ func hasNonblockingTaint(node *v1.Node, nonblockingTaints string) bool {
 	return false
 }
 
+func GetNodeHeartbeatTime(node *v1.Node) metav1.Time {
```
Comment (pohly) [inline test/e2e/framework/node/resource.go:498]: Documentation?

File: test/e2e/framework/node/wait.go (hunk 5 above)
Comment (huww98) [reply test/e2e/framework/node/wait.go:164]: I think `name` is enough, given the function name, package name as the context. `name` is also consistent with other functions in this package.

I've added docs to explain `after`. I think this should be enough. The type `metav1.Time` can also provide reader some context.

File: test/e2e/framework/node/wait.go [hunk 7]
```diff
@@ -160,32 +161,43 @@ func WaitForNodeSchedulable(ctx context.Context, c clientset.Interface, name str
...
 func CheckReady(ctx context.Context, c clientset.Interface, size int, timeout time.Duration) ([]v1.Node, error) {
-	for start := time.Now(); time.Since(start) < timeout; time.Sleep(sleepTime) {
-		nodes, err := waitListSchedulableNodes(ctx, c)
-		if err != nil {
-			framework.Logf("Failed to list nodes: %v", err)
-			continue
-		}
-		numNodes := len(nodes.Items)
+	var nodesAfter []v1.Node
+	gomega.Eventually(ctx, func(g gomega.Gomega) {
```
Comment (pohly) [inline test/e2e/framework/node/wait.go:185]: This looks like a worthwhile improvement, but not one which is related to the original purpose of the PR?

Can you take it out and submit it separately?

Comment (huww98) [reply test/e2e/framework/node/wait.go]: OK, reverted this function

**Review by pohly — COMMENTED** [review]
Looks good to me overall, but I am not familiar with `KubeletCommand` and thus cannot tell whether the proposed changes are okay.

Perhaps someone who has used this and/or written it can chime in?

**Comment by huww98** [comment]
/cc @copejon
As the author of #44923 , can you take a look?

**Comment by pohly** [comment]
/remove-lifecycle stale
/lgtm
/approve

Looks like @copejon missed the ping and I don't know who else to ask for a second opinion, so let's merge it. It's early in the 1.32 development cycle, so if this causes problems, we have time to reconsider.
